File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {
    
    int zero = 0;
    int one = 0;
    public int[] solution(int[][] arr) {
        
        quad(arr, 0, 0, arr.length);
        int[] answer = {zero, one};
        return answer;
    }
    void quad(int[][] arr, int x, int y, int size) {
        if(checkArr(arr, x, y, size)) {
            if(arr[x][y] == 0) zero++;
            else one++;
        }
        else {
            size /= 2;
            quad(arr, x, y, size);
            quad(arr, x+size, y, size);
            quad(arr, x, y+size, size);
            quad(arr, x+size, y+size, size);
        }   
    }
    boolean checkArr(int[][]arr , int x, int y, int size) {
        int n = arr[x][y];
        for(int i = x; i < x + size; i++) {
            for(int j = y; j < y + size; j++) {
                if(n != arr[i][j]) return false;
            }
        }
        return true;
    }
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.js
function isRight(x,y,size,arr) {
    const value = arr[y][x];
    for (let i = y; i < y + size; i++) {
        for (let j = x; j < x + size; j++) {
            if (value !== arr[i][j]) return false;
        }
    }
    return true;
}

function compress(x,y,size,arr,answer) {
    if(isRight(x,y,size,arr)) {
        answer[arr[y][x]]++;
        return 
    }
    let half = size / 2;
    compress(x,y,half,arr,answer);
    compress(x+half,y,half,arr,answer);
    compress(x,y+half,half,arr,answer);
    compress(x+half,y+half,half,arr,answer);
}

function solution(arr) {
    var answer = [0,0];
    compress(0,0,arr.length,arr,answer);
    return answer;
}